### PR TITLE
initial work on supporting dynamic heights

### PR DESCRIPTION
**Pull request by tbranyen**

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -190,12 +222,21 @@ export default class HyperList {
     }
   }
 
+  [_calculatePositions](startFrom) {
+    const itemHeights = this[_itemHeights];
+    const itemPos = this[_itemPositions];
+
+    itemHeights.forEach((itemHeight, i) => {
```
Comment (soyuka) [inline lib/index.js]: shouldn't this use `startFrom`?

File: lib/index.js [hunk 2]
```diff
@@ -120,21 +131,42 @@ export default class HyperList {
...
+    }
+    else if (isNumber(config.itemHeight)) {
+      this[_itemHeights] = Array(config.total).fill(config.itemHeight);
+    }
+    else {
+      this[_itemHeights] = config.itemHeight;
+    }
+
+    // Reuse the item positions if refreshed, otherwise set to empty array.
+    this[_itemPositions] = this[_itemPositions] || [];
```
Comment (soyuka) [inline lib/index.js]: `['foo']` evaluates to false here, it should be `this[_itemPositions] === undefined ? [] : this[_itemPositions]`

File: lib/index.js [hunk 3]
```diff
@@ -239,6 +271,11 @@ export default class HyperList {
     const estTo = from + this[_cachedItemsLen];
     const to = estTo > total ? total : estTo < 0 ? 0 : estTo;
 
+    const fastCache = [];
```
Comment (soyuka) [inline lib/index.js]: `from` is wrong here because `this[_screenItemsLen]` doesn't account the dynamic height.

**Comment by soyuka** [comment]
I left some comments on this would be really neat if you could check it :). Nice work on the library!

File: lib/index.js (hunk 2 above)
Comment (tbranyen) [inline lib/index.js]: Not sure I understand this point, for example:

``` js
const thing = {
  _itemPositions: false, // or any falsy value: 0, undefined, null, NaN, etc...
};

const falsyCase = thing['_itemPositions'] || []; // equals []

thing['_itemPositions'] = [1,2,3];

const truthyCase = thing['_itemPositions'] || []; // equals [1,2,3]

console.log(falsyCase, truthyCase);
// [], [1,2,3]
```

**Comment by tbranyen** [comment]
Thanks so much for the review @soyuka, very much appreciate it! I'll take your changes into account here, but I noticed in your other comment in #12 that you've created a similar project. Would you be interested in folding in your ideas into this project instead of having two competing libs? You can feel free to push directly into this branch (as I given you collab rights).

I'll try and get to the changes you've mentioned today (or tomorrow) otherwise.

File: lib/index.js (hunk 2 above)
Comment (soyuka) [inline lib/index.js]: Hmm yes, my argument is false thanks. Though, any truthy value would be taken and `=== undefined` is cleaner imo (and faster).

**Comment by soyuka** [comment]
Yeah sure, I'll probably have time this weekend to hack things around. WDYT about the position prediction I mentioned in #12?

File: lib/index.js [hunk 4]
```diff
@@ -173,12 +198,20 @@ export default class HyperList {
     }
   }
 
+  _calculatePositions (startFrom) {
+    const itemHeights = this._itemHeights
+    const itemPos = this._itemPositions
+
+    itemHeights.slice(startFrom).forEach((itemHeight, i) => {
```
Comment (soyuka) [inline lib/index.js]: Can you just do a `for (let i = startFrom; i < total; i++)`? `slice` + `forEach` is way slower.

File: lib/index.js [hunk 5]
```diff
@@ -221,6 +245,9 @@ export default class HyperList {
     const from = estFrom > total ? total : estFrom < 0 ? 0 : estFrom
     const estTo = from + this._cachedItemsLen
     const to = estTo > total ? total : estTo < 0 ? 0 : estTo
+    const fastCache = []
```
Comment (soyuka) [inline lib/index.js]: I'll refactor this part soon, I'll open another PR based on your branch in case we disagree on some changes :).

**Pull request closed without merging**